recipes/default.rb

```diff
@@ -34,26 +34,91 @@
   not_if { node["kagent"]["enabled"] == "false" }
 end
 
-# Generate an API key 
-api_key = ""
+# Generate an API key
+ruby_block 'fetch-api-key' do
+  block do
+    require 'net/https'
+    require 'http-cookie'
+    require 'json'
 
-# Template the configuration file 
+    hopsworks_fqdn = consul_helper.get_service_fqdn("hopsworks.glassfish")
+    hopsworks_endpoint = "https://#{hopsworks_fqdn}:#{node['hopsworks']['internal']['port']}"
+    url = URI.parse("#{hopsworks_endpoint}/hopsworks-api/api/auth/service")
+    api_key_url = URI.parse("#{hopsworks_endpoint}/hopsworks-api/api/users/apiKey")
+
+    params =  {
+      :email => node['onlinefs']['hopsworks']['email'],
+      :password => node['onlinefs']['hopsworks']["password"]
+    }
+
+    api_key_params = {
+      :name => "onlinefs",
+      # check with fabio which scopes are needed
+      :scope => "FEATURESTORE,KAFKA,PROJECT"
```

Comment:
If you're uploading data, you will also need the DATASET scopes. If you are running a JOB, you will need the JOB scopes.

Reply:
It looks like you're generating an API key. What happens if you have already generated it and this recipe is being re-run? Recipes must be idempotent. Normally, you would save the API key in a file (read protected by a certain user), and add a guard to this whole ruby block

NOT_IF { ::File::exists?("#{Chef::Config['file_cache_path']}/onlinefs-api-key.txt") }

You would have to save the API key to the file after downloading for this to work:
"#{Chef::Config['file_cache_path']}/onlinefs-api-key.txt")

It means you can skip the blog if the API key already has been downloaded.

Reply:
this api key is not used to write any data. It is only used by the service to get the feature group schema from the kafka schema registry.

About idempotence: So the api key gets saved in the onlinefs-site.xml which gets templated here. So rerunning the recipe I assume the configuration gets newly templated and the old key gets overwritten.

Reply:
talked to fabio, we will change it to read the api-key from a file instead of hardcoding it in the config file. Should also make it easier for the cloud team to replace it.